workloads/ping-pong/ping-pong-client/main.go

```diff
@@ -54,31 +100,100 @@ func getEnv() *Env {
 	return &Env{
 		ServerAddress:    getEnvWithDefault("PING_PONG_SERVICE_HOST", "ping-pong-server.demo"),
 		ServerPort:       getEnvIntWithDefault("PING_PONG_SERVICE_PORT", 8443),
+		MetricsPort:      getEnvWithDefault("METRICS_PORT", ":8080"),
 		SpiffeSocketPath: getEnvWithDefault("SPIFFE_ENDPOINT_SOCKET", "unix:///spiffe-workload-api/spire-agent.sock"),
+		MetricsEnabled:   getEnvBooleanWithDefault("METRICS_ENABLED", true),
+	}
+}
+
+func getEnvBooleanWithDefault(variable string, defaultValue bool) bool {
+	v, ok := os.LookupEnv(variable)
+	if !ok {
+		return defaultValue
+	}
+	switch v {
+	case "true":
+		return true
+	case "false":
+		return false
 	}
+	log.Printf("Invalid value for %s, using default: %v", variable, defaultValue)
+	return defaultValue
 }
 
 func run(ctx context.Context, env *Env) error {
-	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	source, err := workloadapi.NewX509Source(ctx, workloadapi.WithClientOptions(workloadapi.WithAddr(env.SpiffeSocketPath)))
+	// Create X509Source with a separate context for initialization
+	initCtx, initCancel := context.WithTimeout(ctx, 30*time.Second)
+	defer initCancel()
+
+	source, err := workloadapi.NewX509Source(initCtx, workloadapi.WithClientOptions(workloadapi.WithAddr(env.SpiffeSocketPath)))
 	if err != nil {
+		svidFailures.Inc()
 		return fmt.Errorf("unable to obtain SVID: %w", err)
 	}
 	defer source.Close()
 
+	if env.MetricsEnabled {
+		// Expose metrics endpoint
+		http.Handle("/metrics", promhttp.Handler())
+		go func() {
+			if err := http.ListenAndServe(env.MetricsPort, nil); err != nil {
+				log.Printf("Error starting metrics server: %v", err)
+			}
+		}()
+
+		// Monitor SVID updates
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-source.Updated():
+					lastX509SourceUpdate.Set(float64(time.Now().Unix()))
+
+					svid, err := source.GetX509SVID()
+					if err != nil {
+						log.Printf("Error getting X509SVID: %v", err)
+						continue
+					}
+					if len(svid.Certificates) > 0 {
+						notAfter := svid.Certificates[0].NotAfter
+						svidNotAfter.Set(float64(notAfter.Unix()))
+					}
+					// Set the SPIFFE ID URI SAN metric
+					svidURISAN.WithLabelValues(svid.ID.String()).Set(1)
+				}
+			}
+		}()
+
+		// Set initial X509 info in metrics
+		lastX509SourceUpdate.Set(float64(time.Now().Unix()))
+		if svid, err := source.GetX509SVID(); err == nil && len(svid.Certificates) > 0 {
+			svidNotAfter.Set(float64(svid.Certificates[0].NotAfter.Unix()))
+			svidURISAN.WithLabelValues(svid.ID.String()).Set(1)
+		}
+	}
+
 	tlsConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
 	client := &http.Client{
 		Transport: &http.Transport{
 			TLSClientConfig: tlsConfig,
 		},
 	}
 
+	log.Printf("Client starting with metrics at /metrics on %s", env.MetricsPort)
```

Comment:
nit: only true if metrics are enabled. Maybe have a separate "Starting metrics server..." log in the earlier block?